platform/src/main/java/org/hillview/storage/CassandraSSTableLoader.java

```diff
@@ -183,122 +183,83 @@ public Schema getSchema(CFMetaData metadata) {
-                                    cellType = cell.column().cellValueType();
-                                    value = cellType.getSerializer().deserialize(cell.value());
-                                }else{
-                                    /* Hillview won't process the complex data */
-                                    throw new RuntimeException("[Hillview can't convert complex data]");
-                                }
-                                line[i] = value; 
-                            }
-                            i++;
-                        }
-                        /* Append each line to Hillview column */
-                        for (this.currentColumn = 0; this.currentColumn < line.length; this.currentColumn++) {
-                            columns[this.currentColumn].append(line[this.currentColumn]);
-                        }
-                    } else if (unfiltered instanceof RangeTombstoneMarker) {
-                        /* Hillview ignores Tombstone Markers */
-                    }
-                }
-            });
 
-            List<IColumn> arrColumns = new ArrayList<IColumn>(this.sizeToLoad);
-            IMembershipSet membershipSet = null;
-            IAppendableColumn appendableColumn;
-            IColumn column;
-            for (int idx = 0; idx < columns.length; idx++) {
-                appendableColumn = columns[idx];
-                column = appendableColumn.seal();
-                if (membershipSet == null)
-                    membershipSet = new FullMembershipSet(column.sizeInRows());
-                if(columnToLoad[idx])
-                    arrColumns.add(column);
-            }
-            return arrColumns;
+            return CassandraSSTableLoader.this.loadColumns(columnToLoad, partitions, this.sizeToLoad);
         }
     }
 
     public int getNumRows() {
         ISSTableScanner currentScanner = this.ssTableReader.getScanner();
-        Spliterator<UnfilteredRowIterator> splititer = Spliterators.spliteratorUnknownSize(currentScanner, Spliterator.IMMUTABLE);
+        Spliterator<UnfilteredRowIterator> splititer = Spliterators.spliteratorUnknownSize(currentScanner, 
+            Spliterator.IMMUTABLE);
         Stream<UnfilteredRowIterator> partitions = StreamSupport.stream(splititer, false);
-        /* Iterating each item inside the table */
+        // Iterating each item inside the table
         partitions.forEach(partition -> {
             Unfiltered unfiltered;
-            /* Serializing each partition */
-            while (partition.hasNext()){
+            // Serializing each partition
+            while (partition.hasNext()) {
                 unfiltered = partition.next();
-                /* Hillview ignores Tombstone Markers, just count the Row */
-                if (unfiltered instanceof Row){
+                // Hillview ignores Tombstone Markers, just count the Row
+                if (unfiltered instanceof Row) {
```

Comment:
I will use partitions.count(). However, this value might not be accurate since a partition not always contains a valid row, it could be a tombstone marker

Reply:
doesn't the partition maintain a tombstone count then?

Reply:
There is no high-level API, but I can trace how Cassandra gets the table stats. The tracing result will include an estimated number of rows and tombstone counts

Reply:
In principle I think the row count to be exact, I don't know what would happen if we return an approximation.